### PR TITLE
chore: add typed wrappers around State, Reader and Writer

### PR DESCRIPTION
**Pull request by DmitriyMV**
Add typed wrappers around cosi State, Reader and Writer methods and use them.

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: pkg/safe/state.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// Iterator is a generic iterator over resource.Resource slice.
+type Iterator[T any] struct {
+	res []resource.Resource
+	pos int
+}
+
+// NewIterator returns a new iterator over the given slice.
+func NewIterator[T any](res []resource.Resource) Iterator[T] {
```
Comment (smira) [inline pkg/safe/state.go:92]: return from `List()` a typed resource slice, and make `Iterator`

File: pkg/safe/state.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	return Iterator[T]{pos: 0, res: res}
+}
+
+// Next returns the next element of the iterator.
```
[the remark this answers is not in the document]
Comment (DmitriyMV) [reply pkg/safe/state.go:121]: Added new `List[T any]` type which has `Len` and `Get` methods.

File: pkg/safe/state.go (hunk 1 above)
Comment (DmitriyMV) [reply pkg/safe/state.go]: Done.

File: pkg/safe/state.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+)
+
+// StateGet is a type safe wrapper around state.Get.
+func StateGet[T resource.Resource](ctx context.Context, st state.State, ptr resource.Pointer, options ...state.GetOption) (T, error) { //nolint:ireturn
```
[the remark this answers is not in the document]
Comment (DmitriyMV) [reply pkg/safe/state.go:16]: Done.

**Pull request merged**